modules/FiniteSetsExt.tla

```diff
@@ -134,4 +134,43 @@ Choices(Sets) == LET ChoiceFunction(Ts) == { f \in [Ts -> UNION Ts] :
 ChooseUnique(S, P(_)) == CHOOSE x \in S :
                               P(x) /\ \A y \in S : P(y) => y = x
 
+-----------------------------------------------------------------------------
+LOCAL REL == INSTANCE Relation
+
+IsReflexive(op(_,_), S) ==
+    LET R == [ <<x,y>> \in (S \X S) |-> op(x,y) ]
+    IN REL!IsReflexive(R, S)
+
+IsIrreflexive(op(_,_), S) ==
+    LET R == [ <<x,y>> \in (S \X S) |-> op(x,y) ]
+    IN REL!IsIrreflexive(R, S)
+
+IsSymmetric(op(_,_), S) ==
+    LET R == [ <<x,y>> \in (S \X S) |-> op(x,y) ]
+    IN REL!IsSymmetric(R, S)
+
+IsAsymmetric(op(_,_), S) ==
+    LET R == [ <<x,y>> \in (S \X S) |-> op(x,y) ]
+    IN REL!IsAsymmetric(R, S)
+
+IsTransitive(op(_,_), S) ==
+    LET R == [ <<x,y>> \in (S \X S) |-> op(x,y) ]
+    IN REL!IsTransitive(R, S)
+
+IsComparable(op(_,_), S) ==
+    LET R == [ <<x,y>> \in (S \X S) |-> op(x,y) ]
+    IN REL!IsComparable(R, S)
+
+IsPartiallyOrdered(op(_,_), S) ==
+    /\ IsReflexive(op, S)
+    /\ IsTransitive(op, S)
+    /\ IsAsymmetric(op, S)
+
+(***************************************************************************)
+(* Given the binary operator Op, is the set S totally ordered?             *)
+(***************************************************************************)
+IsTotallyOrdered(op(_,_), S) ==
+    /\ IsPartiallyOrdered(op, S)
+    /\ IsComparable(op, S)
+
```

Comment:
The definitions all look reasonable to me. But why do they appear in module FiniteSetsExt? They do not intrinsically require the set S to be finite, except of course that TLC wouldn't be able to evaluate them over infinite sets. The same can be said for module Relation, which requires finiteness only due to the way the transitive closure is defined there. A cleaner structure (but requiring more changes) would perhaps be to move the above operators to Relation.tla, move TransitiveClosure and the subsequent operators to a new module (perhaps FiniteGraph.tla), and drop the instance FiniteSets from Relation?

Reply:
Sure, that sounds like a good idea!

Reply:
I ended up adding new operators to Relation.